Optimize perf for calling ops with custom classes

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38257 Optimize perf for calling ops with custom classes**

It seems we're doing a runtime type check for custom classes on each operator call if the operator has custom class arguments.
This does not have an effect on operators without custom class arguments, but this is a problem for operators with custom class arguments,
for example operators taking a at::native::xnnpack::Conv2dOpContext argument.

The long term solution would be to move those checks to op registration time instead of doing them at call time,
but as an intermediate fix, we can at least make the check fast by

- Using ska::flat_hash_map instead of std::unordered_map
- Using std::type_index instead of std::string (i.e. avoid calling std::hash on a std::string)

Differential Revision: [D21507226](https://our.internmc.facebook.com/intern/diff/D21507226/)

smessmer committed:
Optimize perf for calling ops with custom classes

It seems we're doing a runtime type check for custom classes on each operator call if the operator has custom class arguments.
This does not have an effect on operators without custom class arguments, but this is a problem for operators with custom class arguments,
for example operators taking a at::native::xnnpack::Conv2dOpContext argument.

The long term solution would be to move those checks to op registration time instead of doing them at call time,
but as an intermediate fix, we can at least make the check fast by

- Using ska::flat_hash_map instead of std::unordered_map
- Using std::type_index instead of std::string (i.e. avoid calling std::hash on a std::string)

Differential Revision: [D21507226](https://our.internmc.facebook.com/intern/diff/D21507226/)

[ghstack-poisoned]

smessmer committed:
Update on "Optimize perf for calling ops with custom classes"


It seems we're doing a runtime type check for custom classes on each operator call if the operator has custom class arguments.
This does not have an effect on operators without custom class arguments, but this is a problem for operators with custom class arguments,
for example operators taking a at::native::xnnpack::Conv2dOpContext argument.

The long term solution would be to move those checks to op registration time instead of doing them at call time,
but as an intermediate fix, we can at least make the check fast by

- Using ska::flat_hash_map instead of std::unordered_map
- Using std::type_index instead of std::string (i.e. avoid calling std::hash on a std::string)

Differential Revision: [D21507226](https://our.internmc.facebook.com/intern/diff/D21507226/)

[ghstack-poisoned]

ezyang reviewed (approved):
This seems fine as a short term fix. I agree that we should do a more robust long term fix.

AshkanAliabadi commented:
Removing this check would save us approximately 2.5% of execution time both in quantized and floating point models on mobile.  Thanks for working on this.

smessmer committed:
Update on "Optimize perf for calling ops with custom classes"


It seems we're doing a runtime type check for custom classes on each operator call if the operator has custom class arguments.
This does not have an effect on operators without custom class arguments, but this is a problem for operators with custom class arguments,
for example operators taking a at::native::xnnpack::Conv2dOpContext argument.

The long term solution would be to move those checks to op registration time instead of doing them at call time,
but as an intermediate fix, we can at least make the check fast by

- Using ska::flat_hash_map instead of std::unordered_map
- Using std::type_index instead of std::string (i.e. avoid calling std::hash on a std::string)

Differential Revision: [D21507226](https://our.internmc.facebook.com/intern/diff/D21507226/)

[ghstack-poisoned]
